### ISSUE TITLE
Volumetric Pooling leaving memory in inconsistent state

### ISSUE THREAD

**Issue by Multihuntr** [torch/cunn#284]
The Volumetric Pooling layers (both Max and Average) are causing something to be in an inconsistent state, and thus resulting in a crash. I've not looked deeply enough into the codebase to figure out what, though.

I will describe three ways I can replicate this error. Keeping in mind that they function the same way on at least two different machines with different graphics cards. The same docker environment was used on both machines. 

I have done these tests with VolumetricConvolution and SpatialAveragePooling as well, but have not had this error come up. Tests 1-3 result in the following CUDA error (although, they occur at different places in the nn libraries)

# General Error Message
```
THCudaCheck FAIL file=/root/torch/extra/cutorch/lib/THC/THCGeneral.c line=608 error=9 : invalid configuration argument
/root/torch/install/share/lua/5.1/nn/Container.lua:67: 
In 2 module of nn.Sequential:
/root/torch/install/share/lua/5.1/nn/THNN.lua:109: cuda runtime error (9) : invalid configuration argument at /root/torch/extra/cutorch/lib/THC/THCGeneral.c:608
```

# Test 1 - Most basic example
The following code functions perfectly fine, the output is as expected.
```
require('cunn');
model = nn.Sequential()
model:add(nn.VolumetricMaxPooling(2, 2, 2,  1, 2, 2))
model:add(nn.VolumetricMaxPooling(2, 2, 2,  1, 2, 2))
model:cuda()
someData = torch.CudaTensor(4369, 3, 6, 48, 128) -- Crashes if I change to 4370
print({model:forward(someData)})
```

If I change the 4369 to 4370 and try to run it, I get the following error

```
THCudaCheck FAIL file=/root/torch/extra/cutorch/lib/THC/THCGeneral.c line=608 error=9 : invalid configuration argument
/root/torch/install/share/lua/5.1/nn/Container.lua:67: 
In 2 module of nn.Sequential:
/root/torch/install/share/lua/5.1/nn/THNN.lua:109: cuda runtime error (9) : invalid configuration argument at /root/torch/extra/cutorch/lib/THC/THCGeneral.c:608
stack traceback:
	[C]: in function 'v'
	/root/torch/install/share/lua/5.1/nn/THNN.lua:109: in function 'VolumetricMaxPooling_updateOutput'
	.../torch/install/share/lua/5.1/nn/VolumetricMaxPooling.lua:45: in function <.../torch/install/share/lua/5.1/nn/VolumetricMaxPooling.lua:38>
	[C]: in function 'xpcall'
	/root/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/root/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	[string "print({model:forward(someData)})"]:1: in main chunk
	[C]: in function 'xpcall'
	/root/torch/install/share/lua/5.1/trepl/init.lua:669: in function 'repl'
	/root/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
	[C]: at 0x00406670
```

It uses up around 3GB of memory, but the test has been replicated on three separate machines. Two of those machines have a Titan X in them. All machines used the same docker image with torch on it.

This issue obviously does not occur when using 'nn'.

# Test 2 - Not a threshold value
There is definitely some element of size involved with the crash, but it's not absolute. For example, if I run

```
require('cunn');
model = nn.Sequential()
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:cuda()
someData = torch.CudaTensor(1366, 3, 32, 32, 32)
print({model:forward(someData)})
```

Then it crashes (boundary of crashing at minibatch sizes 1365/1366) with precisely the same error, but only takes up ~700MB.

The precise error is

```
THCudaCheck FAIL file=/root/torch/extra/cutorch/lib/THC/THCGeneral.c line=608 error=9 : invalid configuration argument
/root/torch/install/share/lua/5.1/nn/Container.lua:67: 
In 2 module of nn.Sequential:
/root/torch/install/share/lua/5.1/nn/THNN.lua:109: cuda runtime error (9) : invalid configuration argument at /root/torch/extra/cutorch/lib/THC/THCGeneral.c:608
stack traceback:
	[C]: in function 'v'
	/root/torch/install/share/lua/5.1/nn/THNN.lua:109: in function 'VolumetricAveragePooling_updateOutput'
	...ch/install/share/lua/5.1/nn/VolumetricAveragePooling.lua:20: in function <...ch/install/share/lua/5.1/nn/VolumetricAveragePooling.lua:19>
	[C]: in function 'xpcall'
	/root/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/root/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	[string "print({model:forward(someData)})"]:1: in main chunk
	[C]: in function 'xpcall'
	/root/torch/install/share/lua/5.1/trepl/init.lua:669: in function 'repl'
	/root/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
	[C]: at 0x00406670
```

# Test 3 - Inconsistent state for following operations
This test is a bit strange and I'm not certain how I came up with it; I was just playing around and discovered this. Basically this test shows that it's actually leaving something in an inconsistent state for the next thing to run. Previously, I just saw the error in the Pooling layer because that's all there was.
If you run the following code, it will crash on the tiny allocation of memory for a CudaTensor at the end.

```
require('cunn');
replicateValues = {128, 256}
model2 = nn.Sequential()
for k,v in pairs(replicateValues) do
	model2:add(nn.Select(2, 1)) -- Select the first value from the second dimension
	model2:add(nn.Replicate(v, 1, 4)) -- Replicate v times on the first dimension
	model2:add(nn.VolumetricMaxPooling(1, 1, 1,  1, 2, 2))
end
model2:cuda()
someOtherData = torch.CudaTensor(2, 3, 6, 48, 128)
print({model2:forward(someOtherData)})
torch.CudaTensor(4)
```

This uses ~720MB of space. The precise error is:

```
THCudaCheck FAIL file=/root/torch/extra/cutorch/lib/THC/THCGeneral.c line=608 error=9 : invalid configuration argument
cuda runtime error (9) : invalid configuration argument at /root/torch/extra/cutorch/lib/THC/THCGeneral.c:608
stack traceback:
	[C]: at 0x7fe79fe1f1c0
	[C]: in function 'CudaTensor'
	[string "_RESULT={torch.CudaTensor(4)}"]:1: in main chunk
	[C]: in function 'xpcall'
	/root/torch/install/share/lua/5.1/trepl/init.lua:651: in function 'repl'
	/root/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
	[C]: at 0x00406670
```

# Test 4 - But why does this work?
I tried to prove my hypothesis, but the following code executes without issue, and I have no idea why. I would have expected the separate linear layer to run into an issue, since we were right up against the problem point. So, really, I'm quite lost as to what's happening.

```
require('cunn');
model = nn.Sequential()
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:add(nn.VolumetricAveragePooling(2, 2, 2,  2, 2, 2))
model:cuda()
someData = torch.CudaTensor(1365, 3, 32, 32, 32)
print({model:forward(someData)})
torch.CudaTensor(10000)
```

**Comment by ngimel** [comment]
grid.z dimension in your failing cases is more than the maximum allowed (65535), the launch error is not checked and is caught by subsequent error checks - that's what you are observing. Try using cudnn as a workaround. @soumith, ideally, all kernel launches should be error-checked to avoid propagation of unchecked errors. Error checking right after launch will catch launch errors such as these, but I'm unsure about the clean way of handling errors during kernel execution (e.g. memory access errors) that avoids unnecessary synchronization.

**Comment by soumith** [comment]
closed via #291
